File: Programy/c++/christofides/christofides/christofides.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <algorithm>
#include <cmath>
#include <limits>
#include <set>
#include <queue>
#include <tuple>
#include <utility>
#include <sstream>
#include <chrono>
#include <iomanip>

using namespace std;
using namespace std::chrono;

struct Graph {
    int V;
    std::vector<std::vector<double>> adjMatrix;

    Graph(int V) : V(V), adjMatrix(V, std::vector<double>(V, 0)) {}

    void addEdge(int u, int v, double w) {
        adjMatrix[u][v] = w;
        adjMatrix[v][u] = w;
    }
};

// Funkcja obliczająca MST za pomocą algorytmu Prima
Graph primMST(const Graph& graph, double& mstCost) {
    int V = graph.V;
    Graph mst(V);
    std::vector<double> key(V, std::numeric_limits<double>::max());
    std::vector<int> parent(V, -1);
    std::vector<bool> inMST(V, false);

    key[0] = 0.0;

    for (int count = 0; count < V - 1; ++count) {
        double minKey = std::numeric_limits<double>::max();
        int u;

        for (int v = 0; v < V; ++v) {
            if (!inMST[v] && key[v] < minKey) {
                minKey = key[v];
                u = v;
            }
        }

        inMST[u] = true;

        for (int v = 0; v < V; ++v) {
            if (graph.adjMatrix[u][v] && !inMST[v] && graph.adjMatrix[u][v] < key[v]) {
                parent[v] = u;
                key[v] = graph.adjMatrix[u][v];
            }
        }
    }

    mstCost = 0.0;
    for (int i = 1; i < V; ++i) {
        mst.addEdge(parent[i], i, graph.adjMatrix[parent[i]][i]);
        mstCost += graph.adjMatrix[parent[i]][i];
    }

    return mst;
}

// Funkcja znajdująca wierzchołki o nieparzystym stopniu
std::vector<int> findOddDegreeVertices(const Graph& graph) {
    std::vector<int> oddDegreeVertices;
    for (int i = 0; i < graph.V; ++i) {
        int degree = 0;
        for (int j = 0; j < graph.V; ++j) {
            if (graph.adjMatrix[i][j] != 0) {
                ++degree;
            }
        }
        if (degree % 2 == 1) {
            oddDegreeVertices.push_back(i);
        }
    }
    return oddDegreeVertices;
}

// Funkcja do znajdowania minimalnego doskonałego skojarzenia
std::vector<std::pair<int, int>> minWeightMatching(const Graph& graph, const std::vector<int>& oddDegreeVertices) {
    std::vector<std::pair<int, int>> matching;
    int n = oddDegreeVertices.size();
    std::vector<bool> visited(n, false);

    for (int i = 0; i < n; ++i) {
        if (!visited[i]) {
            int u = oddDegreeVertices[i];
            double minWeight = std::numeric_limits<double>::max();
            int minVertex = -1;

            for (int j = 0; j < n; ++j) {
                if (i != j && !visited[j] && graph.adjMatrix[u][oddDegreeVertices[j]] < minWeight) {
                    minWeight = graph.adjMatrix[u][oddDegreeVertices[j]];
                    minVertex = j;
                }
            }

            visited[minVertex] = true;
            matching.push_back({ u, oddDegreeVertices[minVertex] });
        }
    }

    return matching;
}

std::vector<int> christofidesAlgorithm(const Graph& graph, double& tourCost, double& mstCost) {
    Graph mst = primMST(graph, mstCost);

    std::vector<int> oddDegreeVertices = findOddDegreeVertices(mst);

    std::vector<std::pair<int, int>> matching = minWeightMatching(graph, oddDegreeVertices);

    for (const auto& edge : matching) {
        mst.addEdge(edge.first, edge.second, graph.adjMatrix[edge.first][edge.second]);
    }

    std::vector<int> eulerianCycle;
    std::vector<bool> visited(graph.V, false);
    std::vector<int> stack;
    stack.push_back(0);

    while (!stack.empty()) {
        int u = stack.back();
        bool hasUnvisited = false;

        for (int v = 0; v < graph.V; ++v) {
            if (mst.adjMatrix[u][v] != 0) {
                stack.push_back(v);
                mst.adjMatrix[u][v] = 0;
                mst.adjMatrix[v][u] = 0;
                hasUnvisited = true;
                break;
            }
        }

        if (!hasUnvisited) {
            eulerianCycle.push_back(u);
            stack.pop_back();
        }
    }

    std::vector<int> hamiltonianCycle;
    std::vector<bool> visitedHamiltonian(graph.V, false);

    for (int v : eulerianCycle) {
        if (!visitedHamiltonian[v]) {
            hamiltonianCycle.push_back(v);
            visitedHamiltonian[v] = true;
        }
    }

    tourCost = 0;
    for (size_t i = 0; i < hamiltonianCycle.size() - 1; ++i) {
        tourCost += graph.adjMatrix[hamiltonianCycle[i]][hamiltonianCycle[i + 1]];
    }
    tourCost += graph.adjMatrix[hamiltonianCycle.back()][hamiltonianCycle[0]];

    return hamiltonianCycle;
}

void saveResults(const std::vector<int>& path, double totalCost, double mstCost, double duration) {
    std::ofstream csvFile("results.csv");
    csvFile << "Vertex\n";
    for (int v : path) {
        csvFile << v + 1 << "\n";
    }
    csvFile << "Total cost of the tour," << totalCost << "\n";
    csvFile << "Total cost of the MST," << mstCost << "\n";
    csvFile << std::fixed << std::setprecision(3);
    csvFile << "Execution time (seconds)," << duration << "\n";
    csvFile.close();
}

int main() {
    std::ifstream inputFile("C:\\Users\\mcmys\\OneDrive\\Pulpit\\magisterka repo\\magisterka-repo\\Programy\\testy\\edges100.in");
    if (!inputFile) {
        std::cerr << "Nie można otworzyć pliku." << std::endl;
        return 1;
    }

    int V, E;
    inputFile >> V >> E;
    Graph graph(V);

    int u, v;
    double w;
    while (inputFile >> u >> v >> w) {
        graph.addEdge(u - 1, v - 1, w);
    }

    auto start = high_resolution_clock::now();

    double tourCost, mstCost;
    std::vector<int> tour = christofidesAlgorithm(graph, tourCost, mstCost);

    auto stop = high_resolution_clock::now();
    duration<double> duration = stop - start;

    std::cout << "Znaleziona trasa: ";
    for (int v : tour) {
        std::cout << (v + 1) << " ";
    }
    std::cout << (tour[0] + 1) << std::endl;

    std::cout << "Suma wag krawedzi trasy: " << tourCost << std::endl;
    std::cout << "Suma wag krawedzi w MST: " << mstCost << std::endl;
    std::cout << std::fixed << std::setprecision(3);
    std::cout << "Czas dzialania programu: " << duration.count() << " sekund" << std::endl;

    saveResults(tour, tourCost, mstCost, duration.count());

    return 0;
}

File: Programy/BADANIA/Programy Python/christofides/christofides.py
#!/usr/bin/env python
# coding: utf-8

# In[4]:


import math
import sys
from heapq import heappop, heappush
from itertools import combinations
import time

class Graph:
    def __init__(self, V):
        self.V = V
        self.adjMatrix = [[0] * V for _ in range(V)]

    def add_edge(self, u, v, w):
        self.adjMatrix[u][v] = w
        self.adjMatrix[v][u] = w

def prim_mst(graph):
    V = graph.V
    mst = Graph(V)
    key = [float('inf')] * V
    parent = [-1] * V
    in_mst = [False] * V

    key[0] = 0.0

    for _ in range(V - 1):
        min_key = float('inf')
        u = -1

        for v in range(V):
            if not in_mst[v] and key[v] < min_key:
                min_key = key[v]
                u = v

        in_mst[u] = True

        for v in range(V):
            if graph.adjMatrix[u][v] and not in_mst[v] and graph.adjMatrix[u][v] < key[v]:
                parent[v] = u
                key[v] = graph.adjMatrix[u][v]

    mst_cost = 0.0
    for i in range(1, V):
        mst.add_edge(parent[i], i, graph.adjMatrix[parent[i]][i])
        mst_cost += graph.adjMatrix[parent[i]][i]

    return mst, mst_cost

def find_odd_degree_vertices(graph):
    odd_degree_vertices = []
    for i in range(graph.V):
        degree = sum(1 for j in range(graph.V) if graph.adjMatrix[i][j] != 0)
        if degree % 2 == 1:
            odd_degree_vertices.append(i)
    return odd_degree_vertices

def min_weight_matching(graph, odd_degree_vertices):
    matching = []
    n = len(odd_degree_vertices)
    visited = [False] * n

    for i in range(n):
        if not visited[i]:
            u = odd_degree_vertices[i]
            min_weight = float('inf')
            min_vertex = -1

            for j in range(n):
                if i != j and not visited[j] and graph.adjMatrix[u][odd_degree_vertices[j]] < min_weight:
                    min_weight = graph.adjMatrix[u][odd_degree_vertices[j]]
                    min_vertex = j

            visited[min_vertex] = True
            matching.append((u, odd_degree_vertices[min_vertex]))

    return matching

def christofides_algorithm(graph):
    mst, mst_cost = prim_mst(graph)
    odd_degree_vertices = find_odd_degree_vertices(mst)
    matching = min_weight_matching(graph, odd_degree_vertices)

    for u, v in matching:
        mst.add_edge(u, v, graph.adjMatrix[u][v])

    def find_eulerian_cycle(graph):
        stack = [0]
        eulerian_cycle = []
        adj_matrix_copy = [row[:] for row in graph.adjMatrix]

        while stack:
            u = stack[-1]
            has_unvisited = False

            for v in range(graph.V):
                if adj_matrix_copy[u][v] != 0:
                    stack.append(v)
                    adj_matrix_copy[u][v] = 0
                    adj_matrix_copy[v][u] = 0
                    has_unvisited = True
                    break

            if not has_unvisited:
                eulerian_cycle.append(u)
                stack.pop()

        return eulerian_cycle

    eulerian_cycle = find_eulerian_cycle(mst)
    hamiltonian_cycle = []
    visited = [False] * graph.V

    for v in eulerian_cycle:
        if not visited[v]:
            hamiltonian_cycle.append(v)
            visited[v] = True

    tour_cost = 0
    for i in range(len(hamiltonian_cycle) - 1):
        tour_cost += graph.adjMatrix[hamiltonian_cycle[i]][hamiltonian_cycle[i + 1]]
    tour_cost += graph.adjMatrix[hamiltonian_cycle[-1]][hamiltonian_cycle[0]]

    return hamiltonian_cycle, tour_cost, mst_cost

def save_results(path, total_cost, mst_cost, duration):
    with open("results.txt", "w") as out_file:
        out_file.write("Path: ")
        out_file.write(" ".join(str(v + 1) for v in path))
        out_file.write("\n")
        out_file.write(f"Total cost of the tour: {total_cost}\n")
        out_file.write(f"Total cost of the MST: {mst_cost}\n")
        out_file.write(f"Execution time: {duration} microseconds\n")

def main():
    with open("C:\\Users\\mcmys\\source\\repos\\ConsoleApplication6\\ConsoleApplication6\\edges1000.txt") as input_file:
        V, E = map(int, input_file.readline().split())
        graph = Graph(V)

        for line in input_file:
            u, v, w = map(float, line.split())
            graph.add_edge(int(u) - 1, int(v) - 1, w)

    start = time.time()

    tour, tour_cost, mst_cost = christofides_algorithm(graph)

    duration = (time.time() - start) * 1e6

    print("Znaleziona trasa:", " ".join(str(v + 1) for v in tour), tour[0] + 1)
    print("Suma wag krawedzi trasy:", tour_cost)
    print("Suma wag krawedzi w MST:", mst_cost)
    print("Dlugosc trasy (liczba krawedzi):", len(tour) - 1)
    print("Czas dzialania programu:", duration, "mikrosekund")

    save_results(tour, tour_cost, mst_cost, duration)

if __name__ == "__main__":
    main()


# In[6]:


import matplotlib.pyplot as plt

def read_results(file_path):
    with open(file_path, 'r') as file:
        lines = file.readlines()
        path = list(map(int, lines[0].split()[1:]))  # Odczytaj ścieżkę z 1-bazowymi indeksami
        total_cost = float(lines[1].split()[-1])    # Odczytaj całkowity koszt trasy
        mst_cost = float(lines[2].split()[-1])      # Odczytaj całkowity koszt MST
        duration = float(lines[3].split()[-2])      # Odczytaj czas wykonania jako float
    return path, total_cost, mst_cost, duration

def plot_results(path, total_cost, mst_cost, duration):
    plt.figure(figsize=(10, 6))

    # Tworzenie wykresu
    plt.plot(path + [path[0]], marker='o', linestyle='-', color='b', label='Trasa Hamiltona')
    plt.title('Wizualizacja Trasy Hamiltona - Algorytm Christofidesa')
    plt.xlabel('Wierzchołki')
    plt.ylabel('Kolejność odwiedzin')
    plt.legend()

    # Dodanie tekstu z informacjami
    plt.figtext(0.15, 0.85, f'Suma wag krawędzi trasy: {total_cost}', fontsize=10)
    plt.figtext(0.15, 0.80, f'Suma wag krawędzi w MST: {mst_cost}', fontsize=10)
    plt.figtext(0.15, 0.75, f'Czas działania programu: {duration:.2f} mikrosekund', fontsize=10)

    # Wyświetlenie wykresu
    plt.grid(True)
    plt.savefig('results_visualization.png')  # Zapisanie wykresu do pliku
    plt.show()

def main():
    file_path = 'results.txt'
    path, total_cost, mst_cost, duration = read_results(file_path)
    plot_results(path, total_cost, mst_cost, duration)

if __name__ == "__main__":
    main()


# In[7]:


import networkx as nx
import matplotlib.pyplot as plt

def read_results(file_path):
    with open(file_path, 'r') as file:
        lines = file.readlines()
        path = list(map(int, lines[0].split()[1:]))  # Odczytaj ścieżkę z 1-bazowymi indeksami
        total_cost = float(lines[1].split()[-1])    # Odczytaj całkowity koszt trasy
        mst_cost = float(lines[2].split()[-1])      # Odczytaj całkowity koszt MST
        duration = float(lines[3].split()[-2])      # Odczytaj czas wykonania jako float
    return path, total_cost, mst_cost, duration

def plot_results(path, total_cost, mst_cost, duration):
    G = nx.Graph()

    # Dodanie krawędzi na podstawie ścieżki
    edges = [(path[i], path[i + 1]) for i in range(len(path) - 1)]
    edges.append((path[-1], path[0]))  # Dodanie krawędzi powrotnej do początku
    G.add_edges_from(edges)

    pos = nx.spring_layout(G)

    plt.figure(figsize=(12, 8))
    nx.draw(G, pos, with_labels=True, node_color='skyblue', node_size=500, font_size=10, font_weight='bold', edge_color='gray')

    # Dodanie tytułu i tekstu z informacjami
    plt.title('Wizualizacja Trasy Hamiltona - Algorytm Christofidesa')
    plt.figtext(0.15, 0.85, f'Suma wag krawędzi trasy: {total_cost}', fontsize=10)
    plt.figtext(0.15, 0.80, f'Suma wag krawędzi w MST: {mst_cost}', fontsize=10)
    plt.figtext(0.15, 0.75, f'Czas działania programu: {duration:.2f} mikrosekund', fontsize=10)

    plt.savefig('results_visualization.png')  # Zapisanie wykresu do pliku
    plt.show()

def main():
    file_path = 'results.txt'
    path, total_cost, mst_cost, duration = read_results(file_path)
    plot_results(path, total_cost, mst_cost, duration)

if __name__ == "__main__":
    main()


# In[2]:


import math
import sys
from heapq import heappop, heappush
from itertools import combinations
import time

class Graph:
    def __init__(self, V):
        self.V = V
        self.adjMatrix = [[0] * V for _ in range(V)]

    def add_edge(self, u, v, w):
        self.adjMatrix[u][v] = w
        self.adjMatrix[v][u] = w

def prim_mst(graph):
    V = graph.V
    mst = Graph(V)
    key = [float('inf')] * V
    parent = [-1] * V
    in_mst = [False] * V

    key[0] = 0.0

    for _ in range(V - 1):
        min_key = float('inf')
        u = -1

        for v in range(V):
            if not in_mst[v] and key[v] < min_key:
                min_key = key[v]
                u = v

        in_mst[u] = True

        for v in range(V):
            if graph.adjMatrix[u][v] and not in_mst[v] and graph.adjMatrix[u][v] < key[v]:
                parent[v] = u
                key[v] = graph.adjMatrix[u][v]

    mst_cost = 0.0
    for i in range(1, V):
        mst.add_edge(parent[i], i, graph.adjMatrix[parent[i]][i])
        mst_cost += graph.adjMatrix[parent[i]][i]

    return mst, mst_cost

def find_odd_degree_vertices(graph):
    odd_degree_vertices = []
    for i in range(graph.V):
        degree = sum(1 for j in range(graph.V) if graph.adjMatrix[i][j] != 0)
        if degree % 2 == 1:
            odd_degree_vertices.append(i)
    return odd_degree_vertices

def min_weight_matching(graph, odd_degree_vertices):
    matching = []
    n = len(odd_degree_vertices)
    visited = [False] * n

    for i in range(n):
        if not visited[i]:
            u = odd_degree_vertices[i]
            min_weight = float('inf')
            min_vertex = -1

            for j in range(n):
                if i != j and not visited[j] and graph.adjMatrix[u][odd_degree_vertices[j]] < min_weight:
                    min_weight = graph.adjMatrix[u][odd_degree_vertices[j]]
                    min_vertex = j

            visited[min_vertex] = True
            matching.append((u, odd_degree_vertices[min_vertex]))

    return matching

def christofides_algorithm(graph):
    mst, mst_cost = prim_mst(graph)
    odd_degree_vertices = find_odd_degree_vertices(mst)
    matching = min_weight_matching(graph, odd_degree_vertices)

    for u, v in matching:
        mst.add_edge(u, v, graph.adjMatrix[u][v])

    def find_eulerian_cycle(graph):
        stack = [0]
        eulerian_cycle = []
        adj_matrix_copy = [row[:] for row in graph.adjMatrix]

        while stack:
            u = stack[-1]
            has_unvisited = False

            for v in range(graph.V):
                if adj_matrix_copy[u][v] != 0:
                    stack.append(v)
                    adj_matrix_copy[u][v] = 0
                    adj_matrix_copy[v][u] = 0
                    has_unvisited = True
                    break

            if not has_unvisited:
                eulerian_cycle.append(u)
                stack.pop()

        return eulerian_cycle

    eulerian_cycle = find_eulerian_cycle(mst)
    hamiltonian_cycle = []
    visited = [False] * graph.V

    for v in eulerian_cycle:
        if not visited[v]:
            hamiltonian_cycle.append(v)
            visited[v] = True

    tour_cost = 0
    for i in range(len(hamiltonian_cycle) - 1):
        tour_cost += graph.adjMatrix[hamiltonian_cycle[i]][hamiltonian_cycle[i + 1]]
    tour_cost += graph.adjMatrix[hamiltonian_cycle[-1]][hamiltonian_cycle[0]]

    return hamiltonian_cycle, tour_cost, mst_cost

def save_results(path, total_cost, mst_cost, duration):
    with open("results.txt", "w") as out_file:
        out_file.write("Path: ")
        out_file.write(" ".join(str(v + 1) for v in path))
        out_file.write("\n")
        out_file.write(f"Total cost of the tour: {total_cost}\n")
        out_file.write(f"Total cost of the MST: {mst_cost}\n")
        out_file.write(f"Execution time: {duration:.3f} seconds\n")

    with open("results.csv", "w") as csv_file:
        csv_file.write("Vertex\n")
        for v in path:
            csv_file.write(f"{v + 1}\n")
        csv_file.write(f"Total cost of the tour,{total_cost}\n")
        csv_file.write(f"Total cost of the MST,{mst_cost}\n")
        csv_file.write(f"Execution time (seconds),{duration:.3f}\n")

def main():
    with open("C:\\Users\\mcmys\\OneDrive\\Pulpit\\magisterka repo\\magisterka-repo\\Programy\\testy\\edges100.in") as input_file:
        V, E = map(int, input_file.readline().split())
        graph = Graph(V)

        for line in input_file:
            u, v, w = map(float, line.split())
            graph.add_edge(int(u) - 1, int(v) - 1, w)

    start = time.time()

    tour, tour_cost, mst_cost = christofides_algorithm(graph)

    duration = time.time() - start

    print("Znaleziona trasa:", " ".join(str(v + 1) for v in tour), tour[0] + 1)
    print("Suma wag krawedzi trasy:", tour_cost)
    print("Suma wag krawedzi w MST:", mst_cost)
    print("Czas dzialania programu:", f"{duration:.3f} sekund")

    save_results(tour, tour_cost, mst_cost, duration)

if __name__ == "__main__":
    main()


# In[ ]:




